src/main/java/io/vertx/core/json/impl/JsonUtil.java

```diff
@@ -0,0 +1,114 @@
+ */
+public final class JsonUtil {
+
+  public static final Base64.Encoder BASE64_ENCODER;
+  public static final Base64.Decoder BASE64_DECODER;
+
+  static {
+    /*
+     * Vert.x 3.x Json supports RFC-7493, however the JSON encoder/decoder format was incorrect.
+     * Users who might need to interop with Vert.x 3.x applications should set the system property
+     * {@code vertx.json.base64} to {@code legacy}.
+     */
+    if ("legacy".equalsIgnoreCase(System.getProperty("vertx.json.base64"))) {
+      BASE64_ENCODER = Base64.getEncoder();
+      BASE64_DECODER = Base64.getDecoder();
+    } else {
+      BASE64_ENCODER = Base64.getUrlEncoder().withoutPadding();
+      BASE64_DECODER = Base64.getUrlDecoder();
+    }
+  }
+
+  /**
+   * Wraps well known java types to adhere to the Json expected types.
+   * <ul>
+   *   <li>{@code Map} will be wrapped to {@code JsonObject}</li>
+   *   <li>{@code List} will be wrapped to {@code JsonArray}</li>
+   *   <li>{@code Instant} will be converted to iso date {@code String}</li>
+   *   <li>{@code byte[]} will be converted to base64 {@code String}</li>
+   *   <li>{@code Enum} will be converted to enum name {@code String}</li>
+   * </ul>
+   *
+   * @param val java type
+   * @return wrapped type or {@code val} if not applicable.
+   */
+  public static Object wrapJsonValue(Object val) {
+    if (val == null) {
+      return null;
+    }
+
+    // perform wrapping
+    if (val instanceof Map) {
+      val = new JsonObject((Map) val);
+    } else if (val instanceof List) {
+      val = new JsonArray((List) val);
+    } else if (val instanceof Instant) {
+      val = ISO_INSTANT.format((Instant) val);
+    } else if (val instanceof byte[]) {
+      val = BASE64_ENCODER.encodeToString((byte[]) val);
+    } else if (val instanceof Enum) {
+      val = ((Enum) val).name();
+    }
+
+    return val;
+  }
+
+  @SuppressWarnings("unchecked")
+  public static Object checkAndCopy(Object val) {
+    if (val == null) {
+      // OK
+    } else if (val instanceof Number && !(val instanceof BigDecimal)) {
+      // OK
+    } else if (val instanceof Boolean) {
+      // OK
+    } else if (val instanceof String) {
+      // OK
+    } else if (val instanceof Character) {
+      // OK
+    } else if (val instanceof CharSequence) {
+      val = val.toString();
+    } else if (val instanceof JsonObject) {
+      val = ((JsonObject) val).copy();
+    } else if (val instanceof JsonArray) {
+      val = ((JsonArray) val).copy();
+    } else if (val instanceof Map) {
+      val = (new JsonObject((Map) val)).copy();
+    } else if (val instanceof List) {
+      val = (new JsonArray((List) val)).copy();
+    } else if (val instanceof byte[]) {
+      // OK
+    } else if (val instanceof Instant) {
+      // OK
+    } else if (val instanceof Enum) {
+      // OK
+    } else {
+      throw new IllegalStateException("Illegal type in Json: " + val.getClass());
```

Comment:
@vietj with the goal of being lenient, should we relax this rule?

This code is executed during a copy operation e.g.:

```java
JsonObject other = json.copy()
```

If the goal is to allow everything we might want to allow other types during a copy and only fail if the object is encoded using the codec. wdyt?

This will be a significant change I think that this rule is verified on several tests, but nothing really visible for the end user.